File: python/leetcode/364.py
class Solution(object):
    def depthSumInverse(self, nestedList):
        maxDeep = self.dfsDeep(nestedList)
        return self.dfsHelper(nestedList, 1, maxDeep)

    def dfsDeep(self, nestedList):
        if len(nestedList) == 1 and nestedList[0].isInteger():
            return 1
        deep = 0
        for nInt in nestedList:
            if nInt.isInteger():
                deep = max(deep, 1)
            else:
                deep = max(self.dfsDeep(nInt.getList()) + 1, deep)
        return deep

    def dfsHelper(self, nestedList, level, maxDeep):
        if len(nestedList) == 1 and nestedList[0].isInteger():
            return (maxDeep - level + 1) * nestedList[0].getInteger()
        s = 0
        for nInt in nestedList:
            if nInt.isInteger():
                s += (maxDeep - level + 1) * nInt.getInteger()
            else:
                s += self.dfsHelper(nInt.getList(), level + 1, maxDeep)
        return s

File: cpp/leetcode/364.cpp
class Solution {
public:
    int depthSumInverse(vector<NestedInteger>& nestedList) {
        int res = 0;
        vector<int> v;
        for (auto a : nestedList) helper(a, 0, v);
        for (int i = v.size() - 1; i >= 0; --i) 
            res += v[i] * (v.size() - i);
        return res;
    }
    void helper(NestedInteger ni, int depth, vector<int> &v) {
        if (depth >= v.size()) v.resize(depth + 1);
        if (ni.isInteger()) v[depth] += ni.getInteger();
        else {
            for (auto a : ni.getList()) helper(a, depth + 1, v);
        }
    }
};




class Solution {
public:
    int depthSumInverse(vector<NestedInteger>& nestedList) {
        int unweighted = 0, weighted = 0;
        while (!nestedList.empty()) {
            vector<NestedInteger> nextLevel;
            for (auto a : nestedList) {
                if (a.isInteger()) unweighted += a.getInteger();
                else {
                    nextLevel.insert(nextLevel.end(), a.getList().begin(), a.getList().end());
                }
            }
            weighted += unweighted;
            nestedList = nextLevel;
        }
        return weighted;
    }
};